apps/rush-lib/src/logic/PathTree.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * @public
+ */
+export interface IPathTreeNode<T> {
+  /**
+   * The value that exactly matches the current relative path
+   */
+  value: T | undefined;
+  /**
+   * Child nodes by subfolder
+   */
+  children: Map<string, IPathTreeNode<T>> | undefined;
+}
+
+/**
+ * This class is used to associate POSIX relative paths, such as those returned by `git` commands,
+ * with entities that correspond with ancestor folders, such as Rush Projects
+ */
+export class PathTree<T> {
+  /**
+   * The root node of the tree, corresponding to the path ''
+   */
+  public readonly root: IPathTreeNode<T>;
```

Comment:
Since this is an internal helper, the API design isn't super important.  But here's some design thoughts anyway:  😊

Instead of describing how it's implemented, it might be better to focus on the *point* of this class (which I had some trouble figuring out until I saw the unit test).   Something like this:

```ts
/**
 * Efficiently locates an item that has an associated filesystem path.
 */
class LookupByPath<TItem> {
  /**
   * Adds a new item.  If another item already exists with the same path, an error is reported.
   */
  public addItem(associatedPath: string, item: TItem): void;

  /**
   * Searches for the item with the specified `associatedPath`.
   * @returns the found item, or `undefined` if no item was found
   */
  public findExactPath(associatedPath: string): TItem | undefined;

  /**
   * Searches for the item whose associated path is `childPath` or a parent of that path.
   * If more than one item exists, the closest parent is returned.
   * @returns the found item, or `undefined` if no item was found
   */
  public findChildPath(childPath: string): TItem | undefined;
}

// Example usage:
const rushProjectLookup = new LookupByPath<RushProject>()
rushProjectLookup.addItem('a/b/c', new RushProject());
rushProjectLookup.findExactPath('a/b/c');
rushProjectLookup.findChildPath('a/b/c/d');
```

The tree data structure seems like an implementation detail.  It's probably not even useful for the API to expose it.

Reply:
Adopted some of these changes. Added examples in documentation.